### PR TITLE
GH-40937: [Java] Implement Holder-based functions for ViewVarCharVector & ViewVarBinaryVector

### PR DESCRIPTION
**Pull request by ViggoC**
* GitHub Issue: #40936
* GitHub Issue: #40937

### CODE REVIEW

**Comment by ViggoC** [comment]
@vibhatha @lidavidm Could you review this PR?

File: java/vector/src/main/java/org/apache/arrow/vector/ViewVarCharVector.java [hunk 1]
```diff
@@ -279,6 +313,7 @@ public TransferPair makeTransferPair(ValueVector to) {
   }
 
   private class TransferImpl implements TransferPair {
+
```
Comment (vibhatha) [inline java/vector/src/main/java/org/apache/arrow/vector/ViewVarCharVector.java:316]: Needed?

File: java/vector/src/test/java/org/apache/arrow/vector/TestVarCharViewVector.java [hunk 2]
```diff
@@ -371,6 +373,67 @@ public void testMixedAllocation() {
...
 
+  @Test
+  public void testSetNullableViewVarCharHolder() {
+    try (final ViewVarCharVector viewVarCharVector = new ViewVarCharVector("myvector", allocator)) {
+      viewVarCharVector.allocateNew(0, 0);
+      final List<byte[]> strings = List.of(STR0, STR1, STR2, STR3, STR4, STR5);
+
+      NullableViewVarCharHolder stringHolder = new NullableViewVarCharHolder();
+
+      setAndCheck(viewVarCharVector, 1, strings.get(0), stringHolder);
```
Comment (vibhatha) [inline java/vector/src/test/java/org/apache/arrow/vector/TestVarCharViewVector.java:384]: To check overwrite?

**Review by vibhatha — COMMENTED** [review]
Added a few comments. Thanks for starting this work.

File: java/vector/src/main/java/org/apache/arrow/vector/ViewVarCharVector.java (hunk 1 above)
Comment (ViggoC) [reply java/vector/src/main/java/org/apache/arrow/vector/ViewVarCharVector.java:316]: I guess it's added by my IDE. I'll revert it.

File: java/vector/src/test/java/org/apache/arrow/vector/TestVarCharViewVector.java (hunk 2 above)
Comment (vibhatha) [reply java/vector/src/test/java/org/apache/arrow/vector/TestVarCharViewVector.java:384]: Why do we need this? Could you explain a bit?

Comment (ViggoC) [reply java/vector/src/test/java/org/apache/arrow/vector/TestVarCharViewVector.java:384]: I think it's not needed. It's just missed it while adjusting the code. But you reminded me that it is necessary to add some overwrite test cases.

Comment (vibhatha) [reply java/vector/src/test/java/org/apache/arrow/vector/TestVarCharViewVector.java]: I don't think it is a must. It is okay to leave it at this moment.

File: java/vector/src/main/java/org/apache/arrow/vector/ViewVarCharVector.java [hunk 3]
```diff
@@ -186,9 +213,16 @@ public void setSafe(int index, ViewVarCharHolder holder) {
...
-    // TODO: https://github.com/apache/arrow/issues/40937
-    throw new UnsupportedOperationException(
-        "NullableViewVarCharHolder set operation not supported");
+    if (holder.isSet == 0) {
+      setNull(index);
+    } else {
+      BitVectorHelper.setBit(validityBuffer, index);
+      int start = holder.start;
+      int length = holder.end - start;
+      ArrowBuf dataBuf = holder.buffer;
```
Comment (vibhatha) [inline java/vector/src/main/java/org/apache/arrow/vector/ViewVarCharVector.java:222]: inline this?

File: java/vector/src/main/java/org/apache/arrow/vector/ViewVarCharVector.java [hunk 4]
```diff
@@ -162,8 +185,11 @@ public void get(int index, NullableViewVarCharHolder holder) {
    * @param holder holder that carries data buffer.
    */
   public void set(int index, ViewVarCharHolder holder) {
-    // TODO: https://github.com/apache/arrow/issues/40937
-    throw new UnsupportedOperationException("ViewVarCharHolder set operation not supported");
+    int start = holder.start;
+    int length = holder.end - start;
+    ArrowBuf dataBuf = holder.buffer;
```
Comment (vibhatha) [inline java/vector/src/main/java/org/apache/arrow/vector/ViewVarCharVector.java:190]: inline this?

**Review by vibhatha — APPROVED** [review]
PR LGTM. Just a few minor suggestions. 
Thanks for working on this.

**Comment by vibhatha** [comment]
@lidavidm PR LGTM, please take a look.

**Comment by ViggoC** [comment]
> PR LGTM. Just a few minor suggestions. Thanks for working on this.

Thanks for your review.

**Comment by ViggoC** [comment]
@vibhatha @lidavidm  I think the code to support ViewVarBinaryHolder is nearly duplicate of this PR. Do you prefer to fix it in this one or submit another PR.

**Comment by vibhatha** [comment]
@ViggoC let's do it here.

**Comment by lidavidm** [comment]
Whichever is easier for you.

File: java/vector/src/main/java/org/apache/arrow/vector/ViewVarBinaryVector.java [hunk 5]
```diff
@@ -131,9 +132,32 @@ public byte[] getObject(int index) {
    * @param index position of an element to get
    * @param holder data holder to be populated by this function
    */
-  public void get(int index, NullableViewVarBinaryHolder holder) {
-    // TODO: https://github.com/apache/arrow/issues/40936
-    throw new UnsupportedOperationException("Unsupported operation");
+  public void get(int index, NullableViewVarCharHolder holder) {
```
Comment (vibhatha) [inline java/vector/src/main/java/org/apache/arrow/vector/ViewVarBinaryVector.java:135]: This is incorrect? Does this compile?

**Review by vibhatha — COMMENTED** [review]
Also we need the equivalent tests for binary data.

File: java/vector/src/test/java/org/apache/arrow/vector/TestViewVarBinaryVector.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
+/*
```
Comment (vibhatha) [inline java/vector/src/test/java/org/apache/arrow/vector/TestViewVarBinaryVector.java:1]: I would recommend to keep them in the `TestVarCharViewVector` for the moment. I know the naming is not so matching here. For other `VarBinaryView` tests we used the same test class. And I see it is not very matching. Maybe we could rename the class to `TestVariableWidthViewVector`? This is an issue from my part in the initial PR. We can fix it here.

Comment (ViggoC) [reply java/vector/src/test/java/org/apache/arrow/vector/TestViewVarBinaryVector.java]: Rename the test class to `TestVariableWidthViewVector` and move `testSetNullableViewVarBinaryHolder` to it.

**Review by vibhatha — APPROVED** [review]
Thanks PR LGTM!

**Review by lidavidm — APPROVED** [review]

**Pull request merged**